lib/class-wp-theme-json-gutenberg.php

```diff
@@ -800,13 +800,15 @@ protected static function get_blocks_metadata() {
 		}
 
 		foreach ( $blocks as $block_name => $block_type ) {
-			if (
-				isset( $block_type->supports['__experimentalSelector'] ) &&
-				is_string( $block_type->supports['__experimentalSelector'] )
-			) {
-				static::$blocks_metadata[ $block_name ]['selector'] = $block_type->supports['__experimentalSelector'];
-			} else {
-				static::$blocks_metadata[ $block_name ]['selector'] = '.wp-block-' . str_replace( '/', '-', str_replace( 'core/', '', $block_name ) );
+			$root_selector = static::get_root_block_selector( $block_type );
```

Comment:
Thanks for starting working on this. Got a couple of questions:

**What would you think of extracting the selector logic elsewhere?** I was never a big fan of this logic living in this class. Now that we aim to make it stable, it's a good opportunity to fix it. I'm not opinionated: either a standalone function as in https://github.com/WordPress/gutenberg/pull/45505 or adding a similar method to the `WP_Block_Type` method works for me.

**What are your thoughts on consolidating this API into a single item for consumers?** For example, having a standalone `wp_get_block_css_selector( $block_type, $target = 'root' )` function.

Using this approach, the code here would be refactored to:

```php
		// ...

		foreach ( $blocks as $block_name => $block_type ) {
			// There is always a selector.
			self::$blocks_metadata[ $block_name ]['selector'] = wp_get_block_css_selector( $block_type );

			// The block may or may not have a duotone selector.
			// It could return null for such cases.
			$duotone_selector = wp_get_block_css_selector( $block_type, 'duotone' );
			if ( null !== $duotone_selector ) {
				self::$blocks_metadata[ $block_name ]['duotone'] = $duotone_selector;
			}

			// The block may or may not have a selector for a specific feature.
			// It could return null for such cases.
			foreach( static::BLOCK_SUPPORT_FEATURE_LEVEL_SELECTORS as $key ) {
				$feature_selector = wp_get_block_css_selector( $block_type, $key );
				if ( null !== $feature_selector ) Also: w{
					self::$blocks_metadata[ $block_name ]['features'][ $key ] = $feature_selector;
				}
			}

			// Code stays the same for elements.
		}

		// ...
```

The bit to figure out would be how `wp_get_block_css_selector( $block_type, $target = 'root' )` would work for pulling out selectors for sections (`color`) and individual features (`color.text`). The root one could be the default (`root`).

An alternative could be to have a `wp_get_block_css_selectors( $block_type )` that returns all CSS selectors.

Reply:
Thanks for the suggestion @oandregal 👍 

Extracting the selector logic and providing a simple function for consumers to leverage makes sense to me.

It is proving a little more complicated to achieve than initially thought, but I'm currently working on it and should have something up in the near future. I'll be AFK next week but will continue to chip away at it with the view to push updates the following week.

Reply:
Done in https://github.com/WordPress/gutenberg/pull/46496/commits/323f931a536df2480ced5f790f17a9040047aa0e